mslib/utils/release_info.py

```diff
@@ -0,0 +1,80 @@
+# -*- coding: utf-8 -*-
+"""
+
+    mslib.utils.release_info
+    ~~~~~~~~~~~~~~~~~~~~~~~~
+
+    Shows by a github API call information about the latest release
+
+    This file is part of MSS.
+
+    :copyright: Copyright 2024 Reimar Bauer
+    :copyright: Copyright 2024 by the MSS team, see AUTHORS.
+    :license: APACHE-2.0, see LICENSE for details.
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+"""
+
+import datetime
+import logging
+import requests
+from pebble import concurrent
+
+from mslib.version import __version__ as installed_version
+
+
+@concurrent.process(timeout=1)
+def get_latest_release():
+    # GitHub API URL for the MSS Release
+    url = "https://api.github.com/repos/Open-MSS/MSS/releases/latest"
+
+    try:
+        # Make a GET request to the GitHub API
+        response = requests.get(url, timeout=(1, 1))
+        response.raise_for_status()  # Raise an error for non-200 status codes
+
+        # Extract the JSON response
+        release_data = response.json()
+
+        # Extract the latest release tag and name
+        latest_release = {
+            'tag_name': release_data['tag_name'],
+            'release_name': release_data['name'],
+            'published_at': release_data['published_at'],
+            'url': release_data['html_url']
+        }
+        return latest_release
+
+    except requests.exceptions.RequestException as e:
+        logging.debug(f"Error fetching release data: {e}")
+        return None
+
+
+def check_for_new_release():
+    no_new_release_found = f"{datetime.date.today()}: No new release found."
+    try:
+        # we use the @concurrent.process(timeout=1) on the function to stop requests trying to establish a connection
+        latest_release = get_latest_release().result()
```

Comment:
Immediately calling `.result()` IMO defeats the purpose of using pebble at all. What does this do that isn't already accomplished by just the timeout of `requests.get`?

Reply:
requests.get timeout is used when the connection is established.

But it retrys 10 seconds until it gives up when there is no connection possible.

This is far to long for an user clicking on about, and on the server the startup or restart can be blocked too.

Reply:
Do you have an example to reproduce the issue you see with it? The `requests.get` timeout specifies both the time to wait for a connection to be established and the time to wait for responses from the server. With `timeout=(1, 1)` it will also fail after at most 1 second if there is no connection possible.

The only situation that I am aware of where the request might take longer is when the server is responding very slowly, waiting for almost the timeout before sending another few bytes. That seems to be an unlikely edge case though, especially when accessing GitHub's API?

Reply:
Just load the requests module and disconnect (or have no) internet before you execute

requests.get('https://duckduckgo.com/', timeout=10)

with a connection available this shows a 200, without you wait until you get a requests.exceptions.ConnectionError

We can use MSS offline, we should not add something which can block this.

Reply:
Let's look on the gifcycle how it can  block my system by commenting of the  decorator and starting the local server
![problem](https://github.com/user-attachments/assets/22125c97-76c3-4496-9dce-2b6035f44e73)

Reply:
```
url = "https://api.github.com/repos/Open-MSS/MSS/releases/latest"
requests.get(url, timeout=(1,1))
```
takes 10 sec. without network

Reply:
> Let's look on the gifcycle how it can block my system by commenting of the decorator and starting the local server

It's impossible to see from that gif where that time is actually spent. Can you please measure the execution duration of `get_latest_release()`? Something like this should work:
```diff
diff --git a/mslib/utils/release_info.py b/mslib/utils/release_info.py
index e03ca686..c6809362 100644
--- a/mslib/utils/release_info.py
+++ b/mslib/utils/release_info.py
@@ -33,7 +33,6 @@ from pebble import concurrent
 from mslib.version import __version__ as installed_version
 
 
-@concurrent.process(timeout=1)
 def get_latest_release():
     # GitHub API URL for the MSS Release
     url = "https://api.github.com/repos/Open-MSS/MSS/releases/latest"
@@ -62,12 +61,18 @@ def get_latest_release():
 
 def check_for_new_release():
     no_new_release_found = f"{datetime.date.today()}: No new release found."
+    import time
+    start = time.perf_counter()
     try:
         # we use the @concurrent.process(timeout=1) on the function to stop requests trying to establish a connection
-        latest_release = get_latest_release().result()
+        latest_release = get_latest_release()
     except TimeoutError as e:
         logging.debug(f"Error fetching release data: {e}")
         latest_release = None
+    finally:
+        elapsed = time.perf_counter() - start
+        print("Elapsed time:", elapsed)
+
 
     if latest_release is None or latest_release['tag_name'] == installed_version:
         return no_new_release_found, False
```

Reply:
```
(mssdev) reimarb@infinity:~/PycharmProjects/MSS$ python mslib/mswms/mswms.py 
WARNING:root:No module named 'mswms_settings'. Can't generate gallery.
WARNING:root:Couldn't import mswms_settings (ImportError:'No module named 'mswms_settings''), Using dummy config.
Elapsed time: 10.012056518999998
WARNING:root:Couldn't import mswms_auth (ImportError:'{No module named 'mswms_auth'), creating dummy config.
INFO: Configuration File: 'None'
 * Serving Flask app 'mslib.mswms.app'
 * Debug mode: off
INFO: WARNING: This is a development server. Do not use it in a production deployment. Use a production WSGI server instead.
 * Running on http://127.0.0.1:8081
INFO: Press CTRL+C to quit


```

Reply:
```
url = "https://api.github.com/repos/Open-MSS/MSS/releases/latest"
requests.get(url, timeout=(1,1))
```

my click time is slow ;) but it is now several times repeated and it is ~10secs,

![github_api](https://github.com/user-attachments/assets/0b7c5245-9c4a-4c59-a7d5-34ba2e1ece68)

Reply:
I also recognized, when I have tried the url with network, I need 2 times calling it after the network is off to get to the problem.

Reply:
When we agree to a solution like this, we could also do similiar with the airdata etc. download

Reply:
Hmm, you are getting a different error than I do when disabling the network connection, for some reason. Any idea why? I would consider this a bug in requests, as it clearly does not obey the timeout setting. Could you please report that?

I am still not convinced that pulling in an entire library for threading is the right move to hide a bug behind the heavy weight of setting up an entire thread for a single http request. Where is the thread pool set up for this? How much does setting it up cost for this single usage? Will it be reused if other places also use pebble, or will it need a new one? Wouldn't the python stdlib suffice (there is similar stuff already in `concurrent.futures`)?

Reply:
Which network error do you have? and which requests version, I've just seen I hadn't the  most recent.

Well we change our stack for our needs. When we can replace the other worker usages with it after the GSOC merges it would be usefull too.


I look if that concurrent.futures also helps.

Reply:
With requests                  2.32.3             pyhd8ed1ab_0    conda-forge

it seems to work 
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/reimarb/Miniforge/envs/mssdev/lib/python3.11/site-packages/requests/api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/reimarb/Miniforge/envs/mssdev/lib/python3.11/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/reimarb/Miniforge/envs/mssdev/lib/python3.11/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/reimarb/Miniforge/envs/mssdev/lib/python3.11/site-packages/requests/sessions.py", line 703, in send
    r = adapter.send(request, **kwargs)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/reimarb/Miniforge/envs/mssdev/lib/python3.11/site-packages/requests/adapters.py", line 700, in send
    raise ConnectionError(e, request=request)
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: /repos/Open-MSS/MSS/releases/latest (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x7862fb9f88d0>: Failed to resolve 'api.github.com' ([Errno -3] Temporary failure in name resolution)"))
```

Reply:
mamba list -r
```
2024-09-02 18:09:17  (rev 6)
     certifi  {2024.7.4 (conda-forge/noarch) -> 2024.8.30 (conda-forge/noarch)}
     requests  {2.31.0 (conda-forge/noarch) -> 2.32.3 (conda-forge/noarch)}

2024-09-02 18:16:53  (rev 7)
     requests  {2.32.3 (conda-forge/noarch) -> 2.32.1 (conda-forge/noarch)}

2024-09-02 18:18:25  (rev 8)
     requests  {2.32.1 (conda-forge/noarch) -> 2.31.0 (conda-forge/noarch)}

```
Maybe the problem was only on my system?  It is also gone currently with the old version.

Reply:
> Which network error do you have?

As mentioned above:
```
requests.exceptions.ConnectionError: HTTPSConnectionPool(host='duckduckgo.com', port=443): Max retries exceeded with url: / (Caused by NameResolutionError("<urllib3.connection.HTTPSConnection object at 0x77467b735540>: Failed to resolve 'duckduckgo.com' ([Errno -2] Name or service not known)"))
```

So I am getting a "Errno -2 Name or service not known" while you are getting an "Errno -3 Temporary failure in name resolution".

Reply:
I still can't reproduce the problem I had before.  But I always now have  Errno -3 independent from cable or wireless disconnection